### PR TITLE
Cell model

### PR DESCRIPTION
**Pull request by yann-achard-MS**
Adds a markdown document that motivates and describes the cell model of collaborative editing.

### CODE REVIEW

File: packages/dds/tree/docs/wip/cell-model.md [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+Consider a collaborative editing session between three participants (all acting concurrently), over a field populated with elements [A, B, C, D], where the edits are sequenced as follows:
+
+1. User 1: remove B, C (local state: [A, D])
+
+2. User 2: insert Y after C (local state: [A, B, C, Y, D])
+
+3. User 3: insert X after B (local state: [A, B, X, C, D])
+
+4. User 1: undo the removal of A, B, C (local state: [A, B, C])
```
Comment (scarlettjlee) [inline packages/dds/tree/docs/wip/cell-model.md:23]: I think you mean "undo the removal of B, C (local state: [A, B, C, D])"

Comment (scarlettjlee) [reply packages/dds/tree/docs/wip/cell-model.md:23]: Similarly on line 35.

File: packages/dds/tree/docs/wip/cell-model.md [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+Dynamically-sized fields can be thought of as having list-like behavior: elements can be added anywhere in the list and removed at will which makes the list dynamically grow and shrink.
+Dynamically-sized fields typically start out empty of any cell.
+
+The low-level edits that such fields might support can decomposed as follows:
+
+- Insert: allocate and fill a cell
+
+- Delete/Remove: clear a cell
+
+- Revive: fill a (cleared) cell with the contents it
```
Comment (scarlettjlee) [inline packages/dds/tree/docs/wip/cell-model.md:122]: Should this sentence be completed with "contained before"?

Comment (yann-achard-MS) [reply packages/dds/tree/docs/wip/cell-model.md]: Indeed! Thank you!

File: packages/dds/tree/docs/wip/cell-model.md (hunk 1 above)
Comment (yann-achard-MS) [reply packages/dds/tree/docs/wip/cell-model.md]: Thanks!

File: packages/dds/tree/docs/wip/cell-model.md [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+These additions, while effective, have the following drawbacks:
+
+* They increase the complexity of the core editing model
+- They lead to a set of operations that is not orthogonalized:
+  
+  - Insert, revive, and return seem to share some common traits, with revive and return being closer to each other.
+  
+  - It seems like it should be possible to support a "replace" operation that is similar to revive but with new elements instead of resurrecting removed ones, but that too needs to be added as separate operation)
+
+- They do not alleviate the complexity of modeling an optional field.
```
Comment (Josmithr) [inline packages/dds/tree/docs/wip/cell-model.md:65]: Still relatively new to this problem space, but it isn't immediately clear to me what the complexities here would be. An example or two might help 🙂

File: packages/dds/tree/docs/wip/cell-model.md [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+  
+  - B: allocate and fill a cell with the content being moved
+
+- Return content to A from B:
+  
+  - A: remove the forwarding annotation from a cell and fill it with the returned content
+  
+  - B: clear the (filled) cell
+
+### Edits on Fixed-Sized Fields
```
Comment (Josmithr) [inline packages/dds/tree/docs/wip/cell-model.md:153]: Are the move semantics afforded by dynamically sized fields not useful for fixed-sized arrays? It seems like (unless I'm missing something), a consumer who opted to use a dynamically sized list for some semantically fixed-sized contents would get different merge behaviors than if they used the fixed-sized array.

File: packages/dds/tree/docs/wip/cell-model.md [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+Instead, we see value in offering two sets of higher-level operations: one for fixed-sized fields and one for dynamically-sized fields.
+This allows the editing API to be more specialized for each kind of field, which in turn allows the API to be more expressive and more familiar.
+
+### Merge Semantics
+
+The merge resolution logic can now exclude the possibility of having to merge operations from both sets within a single field (outside of schema migration scenarios).
+
+Since the merge resolution logic is specialized to each field kind, merge resolution logic becomes extensible: we can introduce new kind of fields with associated merge logic.
+This allows us to support new kinds of fields in the future, but more importantly it means that we can migrate from a less desirable set of merge semantics to a more desirable one.
+This dramatically reduces the negative impact of starting out is suboptimal the merge semantics.
```
Comment (Josmithr) [inline packages/dds/tree/docs/wip/cell-model.md:192]: Nit: typo? I didn't follow this.

File: packages/dds/tree/docs/wip/cell-model.md (hunk 4 above)
Comment (yann-achard-MS) [reply packages/dds/tree/docs/wip/cell-model.md:154]: Move semantics are useful for fixed sized fields as well. I updated the document to reflect that.

You're right that a consumer who opted to use a dynamically sized list for some semantically fixed-sized contents would get different merge behaviors than if they used the fixed-sized array.

File: packages/dds/tree/docs/wip/cell-model.md (hunk 3 above)
Comment (yann-achard-MS) [reply packages/dds/tree/docs/wip/cell-model.md]: I added a comment to clarify.

File: packages/dds/tree/docs/wip/cell-model.md (hunk 5 above)
Comment (yann-achard-MS) [reply packages/dds/tree/docs/wip/cell-model.md]: Typo indeed. Let me know if the sentence is clear now.

Comment (Josmithr) [reply packages/dds/tree/docs/wip/cell-model.md]: Yup! That makes sense!

**Review by taylorsw04 — APPROVED** [review]

**Pull request merged**